### PR TITLE
Feat/provider

### PR DESCRIPTION
**Pull request by jasperdg**

### CODE REVIEW

File: oracle/src/data_request.rs [hunk 1]
```diff
@@ -521,9 +530,17 @@ impl Contract {
             payload,
         );
 
+        
         logger::log_new_data_request(&dr);
-
-        self.data_requests.push(&DataRequest::Active(dr));
+        
+        if dr_type == DataRequestType::ProviderPush {
+            let outcome: Outcome = serde_json::from_str(dr.tags[0].as_str()).expect("invalid outcome format");
```
Comment (FranklinWaller) [inline oracle/src/data_request.rs:537]: Using tags for attaching data is a little dirty IMO, but I'm assuming this is temp solution to get this up and running asap

File: oracle/src/data_request.rs [hunk 2]
```diff
@@ -499,15 +502,21 @@ impl Contract {
...
-        self.assert_sender(&config.payment_token);
-        self.dr_validate(&payload);
-        assert!(
-            amount >= validity_bond,
-            "Validity bond of {} not reached, received only {}",
-            validity_bond,
-            amount
-        );
+        let dr_type: DataRequestType = match &payload.provider {
+            Some(account_id) => if account_id == &env::predecessor_account_id() {
```
Comment (FranklinWaller) [inline oracle/src/data_request.rs:506]: Maybe add a comment that says when account_id does not equal to the token address but is a direct call. Took me a bit to get what was going on here

Comment (FranklinWaller) [reply oracle/src/data_request.rs:506]: Btw should this functionality not be whitelisted? Looks like anyone can do a push, or is that desired?

File: oracle/src/data_request.rs [hunk 3]
```diff
@@ -499,15 +502,21 @@ impl Contract {
...
+                DataRequestType::ProviderFetch
+            }, 
+            None => DataRequestType::ValidatorFetch
+        };
+
+        if dr_type == DataRequestType::ValidatorFetch {
+            self.assert_whitelisted(sender.to_string());
+            self.assert_sender(&config.payment_token);
+            self.dr_validate(&payload);
+            assert!(amount >= validity_bond, "Validity bond of {} not reached, received only {}", validity_bond, amount);
```
Comment (FranklinWaller) [inline oracle/src/data_request.rs:518]: Shouldn't the validity bond check be also there if a user wants to use a provider for data fetching?

File: oracle/src/data_request.rs (hunk 2 above)
Comment (jasperdg) [reply oracle/src/data_request.rs:506]: Yes anyone can push

**Review by HeartyMoriarty — APPROVED** [review]

**Pull request closed without merging**